src/main/java/jcifs/smb/SmbTreeConnection.java

```diff
@@ -0,0 +1,892 @@
+                    // immediately
+                    Thread.sleep(500 + RAND.nextInt(1000));
+                }
+            }
+            catch ( InterruptedException e ) {
+                log.debug("interrupted sleep in send", e);
+            }
+
+            if ( request != null ) {
+                log.debug("Restting request");
+                request.reset();
+            }
+            if ( rpath != null ) {
+                // resolveDfs() and tree.send() modify the request packet.
+                // I want to restore it before retrying. request.reset()
+                // restores almost everything that was modified, except the path.
+                rpath.setPath(savedPath);
+                rpath.setFullUNCPath(rpath.getDomain(), rpath.getServer(), savedFullPath);
+            }
+            if ( response != null ) {
+                response.reset();
+            }
+
+            try ( SmbTreeHandle th = connectWrapException(loc) ) {
+                log.debug("Have new tree connection for retry");
+            }
+            catch ( SmbException e ) {
+                log.debug("Failed to connect tree on retry", e);
+                last = e;
+            }
+        }
+
+        if ( last != null ) {
+            log.debug("All attempts have failed, last exception", last);
+            throw last;
+        }
+        throw new SmbException("All attempts failed, but no exception");
+    }
+
+
+    private <T extends CommonServerMessageBlockResponse> T send0 ( SmbResourceLocatorImpl loc, CommonServerMessageBlockRequest request, T response,
+            Set<RequestParam> params ) throws CIFSException, DfsReferral {
+        for ( int limit = 10; limit > 0; limit-- ) {
+            if ( request instanceof RequestWithPath ) {
+                ensureDFSResolved(loc, (RequestWithPath) request);
+            }
+            try ( SmbTreeImpl t = getTree() ) {
+                if ( t == null ) {
+                    throw new CIFSException("Failed to get tree connection");
+                } ;
+                return t.send(request, response, params);
+            }
+            catch ( DfsReferral dre ) {
+                if ( dre.getData().unwrap(DfsReferralDataInternal.class).isResolveHashes() ) {
+                    throw dre;
+                }
+                request.reset();
+                log.trace("send0", dre);
+            }
+        }
+
+        throw new CIFSException("Loop in DFS referrals");
+    }
+
+
+    /**
+     * @param loc
+     * @return tree handle
+     * @throws SmbException
+     */
+    public SmbTreeHandleImpl connectWrapException ( SmbResourceLocatorImpl loc ) throws SmbException {
+        try {
+            return connect(loc);
+        }
+        catch ( UnknownHostException uhe ) {
+            throw new SmbException("Failed to connect to server", uhe);
+        }
+        catch ( SmbException se ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)